tools/codegen/api/python.py

```diff
@@ -0,0 +1,955 @@
+class PythonArgument:
+    name: str
+    type: Type
+    cpp_type_str: str  # TODO: maybe generate on-the-fly?
+    default: Optional[str]
+    default_init: Optional[str]
+
+    # Compute argument formal for python argument parsing.
+    # Needs to be consistent with torch/csrc/utils/python_arg_parser.h.
+    def argument_str(self, *, method: bool = False) -> str:
+        name = self.name
+        typename = _simple_type(self.cpp_type_str)
+
+        # [old codegen] TODO: remove this and make optional types in simple_type
+        # to be consistent across tensor and other types after make Tensor? be
+        # optional instead of undefined
+        if self.type.is_nullable() and '?' not in typename:
+            typename = '{}?'.format(typename)
+
+        # s/self/input/ outside method bindings
+        # [old codegen] TODO: remove this? doesn't rename in codegen, it's just
+        # for the parse string
+        if name == 'self' and typename == 'Tensor' and not method:
+            name = 'input'
+
+        # add list size annotation
+        size = self.size
+        if size is not None:
+            if typename.endswith('?'):
+                typename = f'{typename[:-1]}[{size}]?'
+            else:
+                typename = f'{typename}[{size}]'
+
+        # add default
+        if self.default is not None:
+            default = {
+                'nullptr': 'None',
+                'c10::nullopt': 'None',
+                '{}': 'None',
+            }.get(self.default, self.default)
+            return f'{typename} {name}={default}'
+        else:
+            return f'{typename} {name}'
+
+    @property
+    def size(self) -> Optional[int]:
+        l = self.type.is_list_like()
+        return l.size \
+            if l is not None and l.size is not None and str(l.elem) != 'bool' else None
+
+
+@dataclass(frozen=True)
+class PythonOutArgument(PythonArgument):
+    # In Python signature multiple output fields are packed into one 'out' argument.
+    # When binding to C++, it's first binded to a local 'out' variable:
+    #   'auto out = _r.tensorlist_n<2>(2);',
+    # then binded to scattered C++ output arguments as 'out[0]', 'out[1]', and etc.
+    # TODO: maybe don't need keep scattered out fields for python signature?
+    outputs: Tuple[PythonArgument, ...]
+
+    @staticmethod
+    def from_outputs(outputs: Tuple[PythonArgument, ...]) -> Optional['PythonOutArgument']:
+        if not outputs:
+            return None
+
+        size = len(outputs)
+        if size == 1:
+            return PythonOutArgument(
+                name=outputs[0].name,
+                type=outputs[0].type,
+                cpp_type_str=outputs[0].cpp_type_str,
+                default='None',
+                default_init=None,
+                outputs=outputs,
+            )
+        elif size > 1:
+            return PythonOutArgument(
+                name='out',
+                type=ListType(BaseType(BaseTy.Tensor), size),
```

Comment:
Is it just always the case that ops with more than one output only return tensors?

Reply:
Right - it's enforced in the old codegen (search for "unsupported output argument type" string). I should have added the same assert when creating the python signature!

Reply:
added the assert